[#18] [Backend] As a user, after upload a keywords file, I can see the list of keywords on the screen

suho opened the pull request:
Resolves #18

## What happened 👀

After the user uploads a keywords file, we need to show a list of keywords from that file on the screen.
 
## Insight 📝

- Added Keyword model
- Added pagy for pagination
- Added logic for saving keywords into the database
- Added decorator for showing keyword list in view
 
## Proof Of Work 📹

<img width="1440" alt="Screen Shot 2021-12-03 at 11 24 09" src="https://user-images.githubusercontent.com/19943832/144545215-046472bc-fe5f-4cd8-91fa-eb9cebe6e42f.png">

suho committed:
[#18] Add initializer pagy

suho committed:
[#18] Add keyword model

suho committed:
[#18] Update logic for save keywords

suho committed:
[#18] Implement pagy

suho committed:
[#18] Add decorator to show keyword on the list

suho committed:
[#18] Add empty array for controller

ankitkalia1195 commented on app/views/keywords/index.html.erb:
```diff
@@ -18,41 +18,26 @@
-  </tbody>
-</table>
-
+<% if not @decorated_keywords.empty? %>
```
@suho We should not use instance variables in views as per our compass conventions https://nimblehq.co/compass/development/code-conventions/ruby/ruby-on-rails/#views

suho replied:
Good catch! Thank you! I will update this

suho replied:
Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

ankitkalia1195 commented on db/migrate/20211202054844_create_keywords.rb:
```diff
@@ -0,0 +1,13 @@
+class CreateKeywords < ActiveRecord::Migration[6.1]
+  def change
+    create_table :keywords do |t|
+      t.string :keyword
```
Lets add `null: false` indexes to keyword and status.

suho replied:
Good catch! Thanks! Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

hanam1ni commented on app/decorators/keyword_decorator.rb:
```diff
@@ -0,0 +1,21 @@
+    @keyword = keyword
+  end
+
+  def keyword
```
you can replace this method with `attr_reader`

```ruby
class KeywordDecorator
  attr_reader :keyword

  def initialize(keyword)
    @keyword = keyword
  end

  # ...
end
```

suho replied:
Thanks! Learn about `attr_reader` today! 
Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

ankitkalia1195 commented on app/models/keyword.rb:
```diff
@@ -0,0 +1,11 @@
+
+  enum status: { in_progress: 0, completed: 1, failed: 2 }
+
+  def decorated
```
@suho This will initialize a new object everytime, better to hoist it. 
```
   def decorator
     @decorator ||= KeywordDecorator.new(self)
   end
```

suho replied:
Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

hanam1ni commented on app/decorators/keyword_decorator.rb:
```diff
@@ -0,0 +1,21 @@
+    @keyword.keyword
+  end
+
+  def created_at
```
How about naming this `formatted_created_at`? so it won't replace the original `created_at`

suho replied:
Thanks! Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

hanam1ni commented on db/migrate/20211202054844_create_keywords.rb:
```diff
@@ -0,0 +1,13 @@
+class CreateKeywords < ActiveRecord::Migration[6.1]
+  def change
+    create_table :keywords do |t|
+      t.string :keyword
```
I think we should add null constraint for this column.

suho replied:
Good catch! Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

hanam1ni commented on spec/fabricators/keyword_fabricator.rb:
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+Fabricator(:keyword) do
+  keyword 'MyString'
```
How about using this gem https://github.com/ffaker/ffaker? it can generate random data for test

suho replied:
Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

hanam1ni commented on spec/fabricators/keyword_fabricator.rb:
```diff
@@ -0,0 +1,7 @@
+Fabricator(:keyword) do
+  keyword 'MyString'
+  status ''
+  user nil
```
Should we create user fabricator and assign the user here?

suho replied:
Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

hanam1ni commented on app/views/keywords/index.html.erb:
```diff
@@ -18,41 +18,26 @@
-  </tbody>
-</table>
-
+<% if not @decorated_keywords.empty? %>
```
We can extract this into a partial and pass a keyword to it. This can simplify the view

```ruby
<%= render partial: "keyword_table", locals: {keywords:  @decorated_keywords} %>
```

suho replied:
Learn about partial today 🙈!
Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

junan commented on app/decorators/keyword_decorator.rb:
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class KeywordDecorator
```
This decorator we are using in the view, but based on our [convention](https://nimblehq.co/compass/development/code-conventions/ruby/ruby-on-rails/#decorators), decorators should be used only for the backend, for the view, it should be the presenter. So I think this class should be a [presenter class](https://nimblehq.co/compass/development/code-conventions/ruby/ruby-on-rails/#presenters)

![Screen Shot 2021-12-03 at 12 56 35 PM](https://user-images.githubusercontent.com/6076762/144553073-c851d185-fc3a-4db4-81ef-4a267f859c2c.png)

suho replied:
Thanks for the detailed comment 🙌. I fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

junan commented on app/controllers/keywords_controller.rb:
```diff
@@ -3,10 +3,17 @@
 
   def create
-    parse_keywords
+    parse_keywords.each do |key|
```
How about to use full name?
```suggestion
    parse_keywords.each do |keyword|
```

suho replied:
Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695) 🙈

junan commented on app/views/keywords/index.html.erb:
```diff
@@ -18,41 +18,26 @@
+    </tr>
+    </thead>
+    <tbody>
+    <% @decorated_keywords.each_with_index { |decorated_keyword, index| %>
```
```suggestion
    <% @decorated_keywords.each_with_index do |decorated_keyword, index| %>
```

suho replied:
Good catch! Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

junan commented on app/views/keywords/index.html.erb:
```diff
@@ -18,41 +18,26 @@
+        <td><%= decorated_keyword.created_at %></td>
+        <td><%= raw decorated_keyword.status_html %></td>
+      </tr>
+    <% } %>
```
```suggestion
    <% end %>
```

suho replied:
Good catch! Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

junan commented on db/migrate/20211202054844_create_keywords.rb:
```diff
@@ -0,0 +1,13 @@
+  def change
+    create_table :keywords do |t|
+      t.string :keyword
+      t.integer :status, default: 0
```
I think we can null constraint for this column too

suho replied:
Nice catch! Fixed in [`f42675d` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/f42675dd1e9b7fdfe0b06e96e5e8709d750dd695)

suho committed:
[#18] Imporve code based on reviews
- Add attr_reader,
- Add null constraint to model
- Use partial view
- Remove decorators and use presenters

junan commented on db/schema.rb:
```diff
@@ -10,12 +10,21 @@
   enable_extension "plpgsql"
 
+  create_table "keywords", force: :cascade do |t|
+    t.string "keyword"
```
We added null constraints, but it has not been reflected here, I think, you need to run the migration again

suho replied:
Thanks! I didn't notice that!
Fixed in [`7631238` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/76312381da15a3af06357cf7e84fd9045d11472c)

junan commented on db/schema.rb:
```diff
@@ -10,12 +10,21 @@
 
+  create_table "keywords", force: :cascade do |t|
+    t.string "keyword"
+    t.integer "status", default: 0
```
Same as the above comment

suho replied:
Fixed in [`7631238` (#40)](https://github.com/suho/google-search-ruby/pull/40/commits/76312381da15a3af06357cf7e84fd9045d11472c)

junan reviewed (approved):
Need to re migrate again, so it updates the `schema.rb` file

suho committed:
[#18] Redo migrate
